### ISSUE TITLE
Bug Related to Gradient Accumulation

### ISSUE THREAD

**Issue by Charles20021201** [yongliang-wu/DFT#27]
Dear Authors:

DFT is a really insteresting work! After running several experiements with ```fsdp_dft_trainer.py```, I noticed a bug related to gradient accumulation across micro batches. 

Specifically,

from L457-L460 
```
for micro_batch in micro_batches:
      # Already calls loss.backward() without dividing by n_micro_batches.
      loss, original_loss = self._compute_loss_and_backward(batch=micro_batch) 
      
     # Normalizes loss
      loss = loss / n_micro_batches
      original_loss = original_loss / n_micro_batches
```

In the latest implementation of VeRL SFT trainer, the code is corrected: https://github.com/volcengine/verl/blob/main/verl/trainer/fsdp_sft_trainer.py

```
# Normalizes loss before BP
def _compute_loss_and_backward(self, batch, do_backward=True, n_micro_batches=1):
      loss = torch.sum(loss) / (valid_token_this_rank + 1e-8) * dp_size

      loss = loss / n_micro_batches  # normalize loss

      if do_backward:
          loss.backward()
```

I wonder whether correcting this bug will have a significant influence on the performance gap between DFT and SFT in the reasoning domain, as both of them are implemented incorrectly in some sense. 

Interesting work again! I am looking forward your reply!

Charles Li.

**Comment by yongliang-wu** [comment]
Hi, sorry for the late reply. 

Using the latest version of verl (0.7.0.dev0), I get a 63.66% acc on MATH500 Benchmark using DFT for fine-tuning , which is close to 64.89% reported in the paper.
